source/fab/steps/compile_fortran.py

```diff
@@ -70,111 +91,212 @@ def run(self, artefact_store, config):
+
+        return CompiledFile(
+            input_fpath=analysed_file.fpath, output_fpath=analysed_file.fpath.with_suffix('.o'),
+            source_hash=analysed_file.file_hash, flags_hash=flags_hash,
+            module_deps_hashes=module_deps_hashes
+        )
+
+    def recompile_check(self, analysed_file: AnalysedFile, flags_hash: int, last_compile: Optional[CompiledFile]):
+
+        # todo: other environmental considerations for the future:
+        #   - env vars, e.g OMPI_FC
+        #   - compiler version
+
+        recompile_reasons = []
+        # first encounter?
+        if not last_compile:
+            recompile_reasons.append(NO_PREVIOUS_RESULT)
+
         else:
-            with Timer() as timer:
-                output_fpath.parent.mkdir(parents=True, exist_ok=True)
-
-                command = self.exe.split()
-                command.extend(self.flags.flags_for_path(
-                    path=analysed_file.fpath,
-                    source_root=self._config.source_root,
-                    project_workspace=self._config.project_workspace))
-                command.extend(os.getenv('FFLAGS', '').split())
-                command.append(str(analysed_file.fpath))
-                command.extend(['-o', str(output_fpath)])
-
-                log_or_dot(logger, 'CompileFortran running command: ' + ' '.join(command))
-                try:
-                    run_command(command)
-                except Exception as err:
-                    return Exception("Error calling compiler:", err)
-
-            send_metric(self.name, str(analysed_file.fpath), {'time_taken': timer.taken, 'start': timer.start})
-
-        return CompiledFile(analysed_file, output_fpath)
+            # source changed?
+            if analysed_file.file_hash != last_compile.source_hash:
+                recompile_reasons.append(SOURCE_CHANGED)
+
+            # flags changed?
+            if flags_hash != last_compile.flags_hash:
+                recompile_reasons.append(FLAGS_CHANGED)
+
+            # have any of the modules on which we depend changed?
+            module_deps_hashes = {mod_dep: self._mod_hashes[mod_dep] for mod_dep in analysed_file.module_deps}
+            if module_deps_hashes != last_compile.module_deps_hashes:
+                recompile_reasons.append(MODULE_DEPENDENCIES_CHANGED)
+
+            # is the object file still there?
+            obj_file = analysed_file.fpath.with_suffix('.o')
+            if not obj_file.exists():
+                recompile_reasons.append(OBJECT_FILE_NOT_PRESENT)
+
+            # are the module files we define still there?
+            build_output = self._config.project_workspace / BUILD_OUTPUT
+            mod_def_files = [build_output / f'{mod}.mod' for mod in analysed_file.module_defs]
+            if not all([mod.exists() for mod in mod_def_files]):
+                recompile_reasons.append(MODULE_FILE_NOT_PRESENT)
+
+        return ", ".join(recompile_reasons)
```

Comment:
I just imagined being the user, _"wait, the flags changed too?!"_ or something. Since we can tell the user everything that's changed, it would seem potentially more useful to do so. Happy to reconsider.

Reply:
That seems fine with the exception that if a file is just missing then being told that it doesn't match hashes and such like isn't very helpful. It's implied by the absence of the file.

Reply:
Fab doesn't try to compile files that don't exist.
Do you mean if a mod is missing we shouldn't tell them it's hash has changed?
We don't do that, we only check _mod dependency hashes_ and _mod definition existence_.